rotkehlchen/exchanges/okx.py

```diff
@@ -0,0 +1,410 @@
+            secret=secret,
+            database=database,
+        )
+        self.passphrase = passphrase
+        self.base_uri = 'https://www.okx.com/'
+        self.msg_aggregator = msg_aggregator
+        self.session.headers.update({
+            'OK-ACCESS-KEY': self.api_key,
+            'OK-ACCESS-PASSPHRASE': self.passphrase,
+        })
+
+    def edit_exchange_credentials(
+            self,
+            api_key: Optional[ApiKey],
+            api_secret: Optional[ApiSecret],
+            passphrase: Optional[str],
+    ) -> bool:
+        changed = super().edit_exchange_credentials(api_key, api_secret, passphrase)
+        if api_key is not None:
+            self.session.headers.update({'OK-ACCESS-KEY': self.api_key})
+        if passphrase is not None:
+            self.session.headers.update({'OK-ACCESS-PASSPHRASE': self.passphrase})
+        return changed
+
+    def _generate_signature(self, timestamp: str, method: str, path: str, body: str) -> str:
+        prehash = timestamp + method.upper() + path + body
+        signature = hmac.new(self.secret, prehash.encode('utf-8'), hashlib.sha256)
+        return base64.b64encode(signature.digest()).decode('utf-8')
+
+    def _api_query(
+        self,
+        endpoint: Literal['currencies', 'balance', 'trades', 'deposits', 'withdrawals'],
+        options: Optional[dict] = None,
+    ) -> dict:
+        method = 'get'
+        options = options.copy() if options else {}
+
+        params = {}
+        if endpoint == 'currencies':
+            path = '/api/v5/asset/currencies'
+        elif endpoint == 'balance':
+            path = '/api/v5/account/balance'
+        elif endpoint == 'trades':
+            path = '/api/v5/trade/orders-history-archive'
+            # supports pagination
+            params.update({
+                'limit': options.get('limit', ''),
+                'after': options.get('after', ''),
+            })
+            params.update({
+                'instType': 'SPOT',
+                'state': 'filled',
+            })
+            if options.get('start_ts'):
+                params['begin'] = str(options['start_ts']) + "000"
+            if options.get('end_ts'):
+                params['end'] = str(options['end_ts']) + "000"
+        elif endpoint == 'deposits':
+            path = '/api/v5/asset/deposit-history'
+            # supports pagination
+            params.update({
+                'limit': options.get('limit', ''),
+                'after': options.get('after', ''),
+            })
+        elif endpoint == 'withdrawals':
+            path = '/api/v5/asset/withdrawal-history'
+            # supports pagination
+            params.update({
+                'limit': options.get('limit', ''),
+                'after': options.get('after', ''),
+            })
+        else:
+            raise AssertionError(f'Unexpected {self.name} endpoint type: {endpoint}')
+
+        if len(params) != 0:
+            path += f'?{urlencode(params)}'
```

Comment:
it would be better to use the functionality in requests itself https://requests.readthedocs.io/en/latest/user/quickstart/#passing-parameters-in-urls

Reply:
I had started by using the `params` parameter, but changed to manually concatenating to path in order to use it to generate the signature. I admit it is a little unintuitive, but thought to use this approach after seeing it in implementations for the other exchanges.

I had also tried to directly use requests' approach of generating the URL, but it does not seem to expose that [function](https://github.com/psf/requests/blob/v2.28.1/requests/models.py#L481) for direct use and modifies the `url` attribute on the `Request` class.

There may be a way to initialize the request, use the generated path for the signature, and then attach that as a header, but it seems to diverge from the convention followed across the codebase.

Maybe this can be addressed in a subsequent PR for all instances?